draft-irtf-cfrg-opaque.md

```diff
@@ -1106,55 +989,48 @@ Likewise, `K3dh` is computed by the server as follows:
-This specification can be extended to the 4-message SIGMA-R protocol as used
-in IKEv2.
-
-### SIGMA protocol messages
-
-OPAQUE with SIGMA-I comprises:
-
-- KE1 = credential_request, nonceU, info, epkU
-- KE2 = credential_response, nonceS, epkS, einfo,
-       Sign(skS; transcript2-), MAC(Km2; idS),
-- KE3 = Sign(skU; transcript3-), MAC(Km3; idU)
+#### OPAQUE-3DH Encryption and Key Confirmation {#hmqv-core}
 
-See explanation of fields in {{hmqv-3dh-protocol-messages}}.
-In addition, for the signed material,
-transcript2- is defined similarly to transcript2, however if transcript2 includes
-information that identifies the user, such information can be eliminated in
-transcript2- (this is advised if signing user's identification information by
-the server is deemed to have adverse privacy consequences).
-Similarly, transcript3- is defined as transcript3 with server identification
-information removed if so desired.
+Clients and servers use keys Km2 and Km3 in computing KE2.mac and KE3.mac,
+respectively. These values are computed as HMAC(mac_key, transcript), where
+mac_key and transcript are as follows:
 
-### SIGMA key derivation
+- KE2.mac: mac_key is Km2 and transcript is the concatenation of KE1 and KE2,
+excluding KE2.mac.
+- KE3.mac: mac_key is Km3 and transcript is the concatenation of KE1 and KE2,
+including KE2.mac.
 
-The key schedule for computing Km2, Km3, Ke2, and SK is the same as
-specified in {{hmqv-key-schedule}}. The HKDF input parameter `info` is
-computed as follows:
+The server applicaton info is encrypted using a technique similar to that used
+for secret credential encryption. Specifically, a one-time-pad is derived from
+Ke2 and then used as input to XOR with the plaintext. In pseudocode, this is
+done as follows:
 
 ~~~
-info = "SIGMA-I keys" || I2OSP(len(nonceU), 2) || nonceU
-                      || I2OSP(len(nonceS), 2) || nonceS
-                      || I2OSP(len(idU), 2) || idU
-                      || I2OSP(len(idS), 2) || idS
+info_pad = HKDF-Expand(Ke2, "encryption pad", len(info2))
+einfo = xor(info_pad, info2)
 ~~~
 
-idU and idS are set according to the same rules described in {#derive-hmqv}.
+### Alternate AKE instantiations
 
-The input parameter `IKM` is `Ksigma`, where `Ksigma` is computed by clients
-as `epkS^eskU` and by servers as `epkU^eskS`.
+It is possible to instantiate OPAQUE with other AKEs, such as HMQV {{HMQV}} and SIGMA-I.
+HMQV is similar to 3DH but varies in its key schedule. SIGMA-I uses digital signatures
+rather than static DH keys for authentication. Specification of these instantiations is
+left to future documents.
+
+OPAQUE may also be instantiated with any post-quantum (PQ) AKE protocol that has the message
+flow above and security properties (KCI resistance and forward secrecy) outlined
+in {{security-considerations}}. This document does not specify such an instantiation.
+Note that such an instantiation is not quantum safe unless the OPRF and data encryption schemes
```

Comment:
The data encryption scheme for the envelope encryption is already PQ, and to me, any mention of "data encryption" outside the key exchange step is outside the scope of this doc.

Reply:
How about the following formulation (it refers to 'application data' explicitly to avoid confusion of what data we are talking about):
Note that such an instantiation is not quantum safe unless the OPRF is quantum safe. However, an OPAQUE instantiation where the AKE is quantum safe, but the OPRF is not, would still ensure the confidentiality of application data encrypted under session_key (or a key derived from it) with a quantum-safe encryption function.

Reply:
Btw, we never say what one does with session_key. We may not want to add a derivation of traffic keys as in TLS but we should comment that deriving such keys would be the typical use although applications may have other uses.

Reply:
Thanks, @hugokraw -- I applied your suggested text, and added a note describing that applications can use `session_key` to derive additional keying material.